### DISCUSSION TITLE
Replacing JaninoEventEvaluator

### DISCUSSION THREAD

**Discussion by stephan-roolvink** [qos-ch/logback#888] [Q&A]
In 1.5.13 the JaninoEventEvaluator was removed, I'm seeing that documentation regarding creating filters has not been updated, but I really would like to change my filters so they work with 1.5.13 and beyond. Is there somebody who can help me out?

Out filter looks like this now
```xml
        <filter class="ch.qos.logback.core.filter.EvaluatorFilter">
            <evaluator>
                <matcher>
                    <Name>requestRejectedException</Name>
                    <regex>RequestRejectedException.*</regex>
                </matcher>
                <expression>requestRejectedException.matches(formattedMessage)</expression>
            </evaluator>
            <OnMismatch>NEUTRAL</OnMismatch>
            <OnMatch>DENY</OnMatch>
        </filter>
```

**Comment by ceki** [comment]
**Update:** An online tool called [Janino expression migrator](https://logback.qos.ch/translator/services/janinoExpressionMigrator.html) was deployed earlier today, January 5th 2025. It should help users migrate their configuration files using `JaninoEventEvaluator`.

In general, it should be fairly easy to migrate the functionality of a `JaninoEventEvaluator` setup to a custom `EventEvaluator`.

I suggest to subclass `EventEvaluatorBase` and implement the `evaluate(ILoggingEvent)` method to return a boolean.

For example, here is an evaluator class approximating your example above.

```java

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.IThrowableProxy;
import ch.qos.logback.core.boolex.EvaluationException;
import ch.qos.logback.core.boolex.EventEvaluatorBase;

public class ExceptionMatchEvaluator extends EventEvaluatorBase<ILoggingEvent> {

    String exceptionClass;
    private boolean start = false;

    public void start() {
        if (exceptionClass == null) {
            addError("The exceptionClass must be set");
            return;
        }
        start = true;
    }

    public void stop() {
        start = false;
    }

    public boolean isStarted() {
        return start;
    }

    @Override
    public boolean evaluate(ILoggingEvent event) throws NullPointerException, EvaluationException {

        IThrowableProxy throwableProxy = event.getThrowableProxy();
        if (throwableProxy == null) {
            return false;
        }
        return throwableProxy.getClassName().equalsIgnoreCase(exceptionClass);
    }

    public String getExceptionClass() {
        return exceptionClass;
    }

    public void setExceptionClass(String exceptionClass) {
        this.exceptionClass = exceptionClass;
    }
}
```
And here is a sample logback.xml configuration file screening for `java. lang. RuntimeException`.

```xml
   <configuration>
      <import class="ch. qos.logback.classic.encoder.PatternLayoutEncoder">
      <import class="ch. qos.logback.core.filter.EvaluatorFilter">
      <import class="ch. qos.logback.classic.boolex.ExceptionMatchEvaluator">
      <import class="ch. qos.logback.core.ConsoleAppender">
 
      <appender name="CONSOLE" class="ConsoleAppender">
          <filter class="EvaluatorFilter">
              <evaluator class="ExceptionMatchEvaluator">
                  <exceptionClass>java. lang. RuntimeException</exceptionClass>
              </ evaluator>
              <OnMismatch>DENY</OnMismatch>
              <OnMatch>NEUTRAL</OnMatch>
          </ filter>
 
          <encoder class="PatternLayoutEncoder">
              <pattern>%-4relative [%thread] %-5level %logger -%kvp -%msg%n</pattern>
          </ encoder>
      </ appender>
 
      <root level="INFO">
          <appender-ref ref="CONSOLE">
      </ root>
  </ configuration>
```

**Comment by ceki** [comment]
Added `ExceptionMatchEvaluator` in commit 020610a44507ab6152eb3e6b6736fe073f633648

Documentation will be updated shortly

**Comment by Cesaario** [comment]
I had some problems with JaninoEventEvaluator after updating `logback` to version `1.5.15` but I couldn't solve it with the solution from above. I was able to solve this issue by updating `logback-access` to version `2.0.5`. Just be aware that you will need to rename the package from `ch.qos.logback.access:common` to `ch.qos.logback.access:logback-access-common` (the same if you use `ch.qos.logback.access:tomcat`).

**Comment by ceki** [comment]
@Cesaario Your comment seems unrelated to the issue at hand, namely the removal of JaninoEventEvaluator. May I suggest to create a new discussion?

**Reply by Cesaario** [reply]
Sorry about that. It's just because I stumbled upon this discussion when searching for a solution for my problem, so maybe it could be useful for someone with the same problem. I don't think a new discussion is needed.

**Comment by ceki** [comment]
@stephan-roolvink  An online tool called [Janino expression migrator](https://logback.qos.ch/translator/services/janinoExpressionMigrator.html) was deployed earlier today, January 5th 2025. It should help users migrate their configuration files using `JaninoEventEvaluator`.

**Comment by stephan-roolvink** [comment]
Thanks for the answer I was aware of creating an evaluator for this, I was hoping for a way to avoid it as it add more code to maintain.
